### ISSUE TITLE
Add Logical Boolean Operations

### ISSUE THREAD

**Issue by adishavit** [cvxgrp/cvxpy#1000]
When solving mixed integer programs which some binary variables, it is often desirable to use logical Boolean operators like OR, AND, NOT, XOR (and also a few more complex ones). 

These can all be expressed as MILP constrains, as shown e.g. [here](https://cs.stackexchange.com/questions/12102/express-boolean-logic-operations-in-zero-one-integer-linear-programming-ilp/43884#43884). These are simple transformation but are amenable to errors, and using operator overloading or named functions for them would simplify and clarified user code. 

I think cvxpy would benefit from such an addition.

**Comment by rileyjmurray** [comment]
I agree that those operators would be convenient. It's also the kind of basic rewriting that cvxpy is designed to accommodate.

There are several design decisions to be made, however. One clear decision is whether such functionality would be exposed by operator overloading, or whether we would use named functions (perhaps in a namespace like ``cp.logicals``). There are also more subtle points, such as how cvxpy verifies validity of inputs to those functions. For example, if ``x`` is a continuous variable and ``y`` is an integer, then we would probably want to raise an error if someone wrote ``cp.logicals.AND(x,y)``.

Due to bandwidth limitations of cvxpy developers, I think we would need another volunteer in order to seriously explore this possibility in the short to medium term. If anyone (including you @adishavit ) would be interested in contributing in such a way, please indicate so on this thread.

**Comment by adishavit** [comment]
I’m not a Python expert (and quite busy) but am certainly willing to learn more about how to contribute. 
I’d like to see a discussion on the design decision here first. I’ve never checked cvxpy internals.
Are the dev docs? Internals docs?

**Comment by adishavit** [comment]
I’m guessing named functions can be added first and any overloaded operators if/when added may just call the names functions. 
I _would_ like to have the option to incrementally build an expression like `x0&&x1&&...&&xn`. Like an `&&=` operator.

**Comment by rileyjmurray** [comment]
@adishavit we're in the process of merging in some significantly improved documentation for contributing to CVXPY (#1020). This particular situation is more like contributing a transform. I contributed a pretty complicated atom + transform about six months ago (#860). In your case I think it would be a much simpler transform. You could define the functions in a file ``cvxpy/transforms/logic.py``, with more or less each function being an independent implementation (no strict need to organize everything under a common class).

**Comment by adishavit** [comment]
Looking at the simpler transforms they seem relatively straight forward though I don’t really understand if they must support any particular API. 
The documentation I saw is very partial and confusing. Does you newer documentation have more details?
Is there a necessary and sufficient API the must be conformed to to create a transform?

**Comment by rileyjmurray** [comment]
There isn't a necessary and sufficient API, actually. From a practical perspective I think anything can go in ``cvxpy/transforms`` unless there's a definitively better place to put it. Right now all discrete constraints are enforced by properties assigned to a Variable object (i.e. there is no``cvxpy/constraints/bool.py`` or ``cvxpy/constraints/integer.py``). So I suppose -- in this particular case, one API requirement is that an end-user can't be provided with a single classical CVXPY Constraint. To be concrete, let's say the API is something like
```
val = cvxpy.transforms.logic.xor(expr1, expr2)
```
In this case the user can't update``constrs.append(val)`` and then call ``prob = Problem(objective, constrs)``, because CVXPY will get to ``val`` when iterating over the Problem's constraints and then not know what to do with ``val``.

One option is for these functions to return lists of CVXPY constraints. So an end-user could do something like
```
# constrs = [...], objective = ...
val = cvxpy.transforms.logic.xor(expr1, expr2)
constrs.extend(val)  # or constrs += val
prob = cvxpy.Problem(objective, constrs)
```
That has the advantage of being pretty simple to implement. However, it has a *disadvantage* of making it hard to interact with the abstraction of ``val`` as an XOR constraint down the line. Do you think that disadvantage would be severe?

**Comment by adishavit** [comment]
I'm not sure I quite see the distinction unless you mean that in the former case `val` is a `Variable` and in the latter an actual constraint.

In any case, what you suggested is not so far from what I already wrote in my own code:

```python
def cast_non_neg_bounded_integer_to_boolean(x: cp.Variable, upper_bound: int):
    """
    From: https://cs.stackexchange.com/a/12118/1813
    Cast to boolean (version 1):
        Suppose you have an integer variable x, and you want to define y so that y=1 if x≠0 and y=0 if x=0.
        If:
            0 ≤ x ≤ U
        then you can use the linear inequalities:
            0 ≤ y ≤ 1,
            y ≤ x,
            x ≤ Uy
        Note: this ONLY works if 0 ≤ x ≤ U.
    """
    y = cp.Variable(1, name=f"Is_{x.name()}_non_zero", integer=True)
    return y, [0 <= x, x <= upper_bound,
               0 <= y, y <= 1, # Would these be implied by specifying y w/ boolean=True
               y <= x,
               x <= upper_bound * y]
```
Here I return both new variable and the additional constraints.
IIUC, this does not have the disadvantage you are referring to. The external code may use and compose `y` into other/larger expressions and extend the constraints with the newly added constraints.

**Comment by giacomoverardo** [comment]
Hi everyone, I have the same issue. Have logical boolean operations been implemented?

**Comment by adishavit** [comment]
I have something along these lines, though I'm not sure where/how to add it to CVXPY:

```python
def logical_and(y: cp.Variable, xs: List[cp.Variable or int or bool]):
    """
    define y to be the conjunction (Logical AND) of all the xs:

        y == x0 ∧ x1 ∧ ... ∧ x2

    When there is just one x0 then y==x0

    See:
    - http://yetanothermathprogrammingconsultant.blogspot.com/2016/02/xor-as-linear-inequalities.html?m=0
    - https://cs.stackexchange.com/a/43884

    :param y: a cp.Variable(1, boolean=True)
    :param xs: List[cp.Variable(1, boolean=True)]
    :returns: the constraints required for getting this relationship
    """
    assert all([x.attributes['boolean'] for x in xs if isinstance(x, cp.Variable)] + [y.attributes['boolean']])
    n = len(xs)
    expr = sum(xs) - n * y
    return [0 <= expr, expr <= n - 1]


# take a list of variables or expressions and return a boolean variable representing their conjunction, i.e. logical AND. 
def and_(xs: List[cp.Variable or int or bool], cstrs: list):
    y = cp.Variable(1, boolean=True)
    cstrs += logical_and(y, xs)
    return y
```
Note that you must pass an in-out list of constraints to the function.   
The operator `or_` could be implemented similarly.
